### PR TITLE
refactor registry into single list 

### PR DESCRIPTION
**Pull request by joshm-3b6b58**
This simplifies the registry list into a single list rather than separate lists for directories and keys. This will make adding features easier down the line.

This makes the code easier to read by getting rid of a bunch of silly flags denoting the type of registry entry, etc.

This is the first in a series of changes I will be making to this. Right now the entries in the list aren't un-selectable, but they will be when I enable the selection of multiple entries.

### CODE REVIEW

File: client-js/gulpfile.js [hunk 1]
```diff
@@ -51,7 +51,8 @@ gulp.task('compile-ts', function () {
       declarationFiles: false,
       noExternalResolve: true,
       experimentalDecorators: true,
-      emitDecoratorMetadata: true
+      emitDecoratorMetadata: true,
+      experimentalAsyncFunctions: true
```
Comment (joshm-3b6b58) [inline client-js/gulpfile.js:55]: I had to add this on my end to silence a whole bunch of compiler warnings...

File: client-js/app/elements/labrad-registry.html [hunk 2]
```diff
@@ -101,9 +101,9 @@
   <template>
     <div id="reg-area">
       <table>
-        <tbody is="selectable-table" attr-for-selected="name" selected="{{selDir}}" on-tap="selectDir">
+        <tbody is="selectable-table" id="combinedList" selected="{{idx}}">
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.html]: maybe rename `idx` to something like `selectedIdx` to be more specific

File: client-js/app/elements/labrad-registry.html [hunk 3]
```diff
@@ -101,9 +101,9 @@
...
           <template id="dirList" is="dom-repeat" items="{{dirs}}" filter="filterFunc">
-            <tr name="{{item.name}}">
+            <tr name="{{item.name}}" class = "dir">
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.html]: nit: remove spaces around `=` for element attributes

File: client-js/app/elements/labrad-registry.html [hunk 4]
```diff
@@ -142,15 +140,15 @@
           <iron-icon icon="folder-open"></iron-icon>
           new dir
         </paper-button>
-        <paper-button on-click="copyClicked" disabled="{{!selected}}">
+        <paper-button on-click="copyClicked" disabled="{{selected}}">
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.html]: Seems strange to change the sense of `selected` here. Shouldn't these buttons be disabled only if nothing is selected?

File: client-js/app/elements/labrad-registry.ts [hunk 5]
```diff
@@ -61,31 +53,49 @@ export class LabradRegistry extends polymer.Base {
     return item.name.match(this.regex);
   }
 
-  selectKey() {
-    this.selDir = null;
-    this.selectType = 'key';
+  @listen('idx-changed')
+  changedSelected() {
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: Is this just here for debugging? Could probably remove since things are working.

File: client-js/app/elements/labrad-registry.ts [hunk 6]
```diff
@@ -61,31 +53,49 @@ export class LabradRegistry extends polymer.Base {
...
+    console.log('ChangeSelected', this.idx);
   }
 
-  selectDir() {
-    this.selKey = null;
-    this.selectType = 'dir';
+  @computed()
+  selected(idx: number): boolean {
+    console.log("computing selected is: ", idx);
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: This log message doesn't make sense to me.

File: client-js/app/elements/labrad-registry.ts [hunk 7]
```diff
@@ -61,31 +53,49 @@ export class LabradRegistry extends polymer.Base {
...
+    if (idx != null) {
+        return false;
+      } else {
+        console.log('doesnt exist');
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: ???

File: client-js/app/elements/labrad-registry.ts [hunk 8]
```diff
@@ -61,31 +53,49 @@ export class LabradRegistry extends polymer.Base {
...
+        return true;
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: This seems backwards. `selected` should be true if idx is non-null:

``` javascript
return idx != null;
```

File: client-js/app/elements/labrad-registry.ts [hunk 9]
```diff
@@ -61,31 +53,49 @@ export class LabradRegistry extends polymer.Base {
...
+      }
+  }
+  @property({computed: '_computeSelectedType(idx)'})
+  selectedType: string;
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: These can be written with the `computed` decorator, just like `selected` above:

``` javascript
@computed()
selectedType(idx: number): string {
  ...
}
```

Same for `selectedItem`.

File: client-js/app/elements/labrad-registry.ts [hunk 10]
```diff
@@ -61,31 +53,49 @@ export class LabradRegistry extends polymer.Base {
...
+  
+  @property({computed: '_computeSelectedItem(idx)'})
+  selectedItem: string;
+  
+  _computeSelectedType(idx: number): string {
+    if (typeof this.dirs !== 'undefined') { //I don't know if there's a better way to do this
+      if (idx < this.dirs.length) {
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: You could combine the two conditions:

``` javascript
if (this.dirs && idx < this.dirs.length) {
  return 'dir';
} else {
  return 'key';
}
```

File: client-js/app/elements/labrad-registry.ts [hunk 11]
```diff
@@ -95,6 +105,7 @@ export class LabradRegistry extends polymer.Base {
     for (var j in resp.keys) {
       this.push('keys', {name: resp.keys[j], value: resp.vals[j]});
     }
+    this.$.combinedList.selected = null;
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: Can you do this by assigning to `this.idx` instead of touching the DOM directly?

Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.ts]: I tried that, it didn't work for me.

File: client-js/app/elements/labrad-registry.ts [hunk 12]
```diff
@@ -366,11 +377,11 @@ export class LabradRegistry extends polymer.Base {
     var newPath = JSON.parse(this.$.copyPathInput.value);
 
     try {
-      if (this.selectType === 'dir') {
-        await this.socket.copyDir({path: this.path, dir: this.selDir, newPath: newPath, newDir: newName});
+      if (this.selectedType == 'dir') {
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: Still best to use triple equals here (and numerous places below). Double equals can do weird type coercions and it's best not to get in the habit of using it.

File: client-js/app/elements/selectable-table.html [hunk 13]
```diff
@@ -15,7 +15,7 @@
       is: 'selectable-table',
       extends: 'tbody',
       behaviors: [
-        Polymer.IronSelectableBehavior
+        Polymer.IronMultiSelectableBehavior
```
Comment (maffoo) [inline client-js/app/elements/selectable-table.html]: Probably should leave this multi-selection for later, since we currently only support single selection.

File: client-js/app/elements/labrad-registry.ts (hunk 9 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.ts]: I tried that first, but the compiler doesn't like it. It seems selectedType as something odd like a function that returns a string, rather than just a string and it throws errors. This was the only way I could work it so the compiler would be happy.

File: client-js/gulpfile.js (hunk 1 above)
Comment (maffoo) [inline client-js/gulpfile.js:55]: This is fine, but maybe should put it in a separate commit when merging.

File: client-js/app/elements/labrad-registry.ts (hunk 9 above)
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: I see, you mean when you refer to `this.selectedType` in the code it picks up the method, rather than the computed property value.

File: client-js/app/elements/labrad-registry.ts [hunk 14]
```diff
@@ -61,31 +53,49 @@ export class LabradRegistry extends polymer.Base {
...
+  selected(idx: number): boolean {
+    console.log("computing selected is: ", idx);
+    if (idx != null) {
+        return false;
+      } else {
+        console.log('doesnt exist');
+        return true;
+      }
+  }
+  @property({computed: '_computeSelectedType(idx)'})
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: nit: add blank line between this decorator and the previous method.

**Comment by joshm-3b6b58** [comment]
Thanks for the comments. Since this is UI heavy I wasn't writing test until we implement Selenium or some kind of browser testing framework. Are there any tests you would add for this?

File: client-js/app/elements/selectable-table.html (hunk 13 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/selectable-table.html]: Thanks for picking this up, I was going to add multi-selectable to this PR, but figured I do this one thing at a time.

File: client-js/app/elements/labrad-registry.html (hunk 2 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.html]: OK

File: client-js/app/elements/labrad-registry.html (hunk 3 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.html]: ah yes. OK

File: client-js/app/elements/labrad-registry.ts (hunk 5 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.ts]: Yup

File: client-js/app/elements/labrad-registry.ts (hunk 6 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.ts]: It's a leftover from some old debugging. Removed it.

File: client-js/app/elements/labrad-registry.ts (hunk 9 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.ts]: Yes, if I put it in the `if` statements on `doCopy()` for example. Also, they're passed to the RPC calls on line 375, for example. Those generate compiler errors.

File: client-js/app/elements/labrad-registry.ts (hunk 8 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.ts]: Yeah, sorry I confused myself when I changed something. I've re-reversed the returns in these if statements now.

File: client-js/app/elements/labrad-registry.ts (hunk 10 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.ts]: Ah thank you, I didn't realize you could do that in js. I knew it worked in python.

File: client-js/app/elements/labrad-registry.ts (hunk 12 above)
Comment (joshm-3b6b58) [inline client-js/app/elements/labrad-registry.ts]: OK

**Comment by joshm-3b6b58** [comment]
@maffoo I think I made all the changes, can you take a look when you get a chance? Thanks.

File: client-js/gulpfile.js (hunk 1 above)
Comment (joshm-3b6b58) [inline client-js/gulpfile.js:55]: I may need a hand doing that when I rebase and squash...

File: client-js/app/elements/labrad-registry.ts [hunk 15]
```diff
@@ -61,31 +53,41 @@ export class LabradRegistry extends polymer.Base {
     return item.name.match(this.regex);
   }
 
-  selectKey() {
-    this.selDir = null;
-    this.selectType = 'key';
+  @computed()
+  selected(selectedIdx: number): boolean {
+    if (selectedIdx != null) {
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: I would suggest just

``` javascript
return selectedIdx !== null;
```

(note use of `!==` instead of `!=`)

File: client-js/app/elements/labrad-registry.ts [hunk 16]
```diff
@@ -61,31 +53,41 @@ export class LabradRegistry extends polymer.Base {
...
+      return 'key';
+    }
   }
 
-  @computed()
-  selected(selectType: string, selDir: string, selKey: string): boolean {
-    console.log('selectType', selectType, 'selDir', selDir, 'selKey', selKey);
-    return (selectType === 'dir' && selDir != null) || (selectType === 'key' && selKey != null);
+  _computeSelectedItem(selectedIdx: number): string {
+    if (typeof this.dirs !== 'undefined') {
```
Comment (maffoo) [inline client-js/app/elements/labrad-registry.ts]: can do the same thing here as in `_computeSelectedType`:

``` javascript
if (this.dirs) {
  return ...
}
```

**Comment by maffoo** [comment]
Two small comments, otherwise LGTM.
